### PR TITLE
add barabasi_albert model

### PR DESCRIPTION
**Pull request by afternone**
```
using LightGraphs
using StatsBase
using PyPlot

julia> @time g = barabasi_albert(100_000, 3)
  0.376023 seconds (2.88 M allocations: 105.213 MB, 9.96% gc time)
{100000, 299991} undirected graph

julia> c = countmap(degree(g))
julia> loglog(collect(keys(c)), collect(values(c)),"o")
julia> xlabel("degree")
julia> ylabel("count")
```
BA network

### CODE REVIEW

File: src/randgraphs.jl [hunk 1]
```diff
@@ -148,6 +148,52 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
...
+are preferentially attached to existing nodes with high degree. Undirected graphs are
+created by default; use `is_directed=true` to override.
+"""
+function barabasi_albert(n::Integer, m::Integer; is_directed=false)
+	@assert(1<=m<n, "Barabási–Albert network must have 1 <= m < n")
+	if is_directed
+		g = DiGraph(n)
+	else
+		g = Graph(n)
+	end
```
Comment (sbromberger) [inline src/randgraphs.jl]: probably better: `g = is_directed ? DiGraph(n) : Graph(n)`

File: src/randgraphs.jl [hunk 2]
```diff
@@ -148,6 +148,52 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
...
+	# Target nodes for new edges
+	targets = collect(1:m)
```
Comment (sbromberger) [inline src/randgraphs.jl]: Do you need to `collect` these or can you use the range? (The range will save memory.)

File: src/randgraphs.jl [hunk 3]
```diff
@@ -148,6 +148,52 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
...
+			target = sample(repeated_nodes)
+			if !node_status[target]
+				targets[i] = target
+				i += 1
+				node_status[target] = true
+			end
+		end
+		source += 1
+	end
+	g
```
Comment (sbromberger) [inline src/randgraphs.jl]: make the `return` explicit here, please.

**Comment by sbromberger** [comment]
Love it. Thanks.

File: src/randgraphs.jl (hunk 2 above)
Comment (sbromberger) [inline src/randgraphs.jl]: ...

File: src/randgraphs.jl [hunk 4]
```diff
@@ -148,6 +148,48 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
...
+"""Creates a [Barabási–Albert model](https://en.wikipedia.org/wiki/Barab%C3%A1si%E2%80%93Albert_model) 
+random graph with `n` nodes is grown by attaching new nodes each with `m` edges that 
+are preferentially attached to existing nodes with high degree. Undirected graphs are
+created by default; use `is_directed=true` to override.
+"""
+function barabasi_albert(n::Integer, m::Integer; is_directed=false)
+	@assert(1<=m<n, "Barabási–Albert network must have 1 <= m < n")
+	g = is_directed ? DiGraph(n) : Graph(n)
+	# Target nodes for new edges
+	targets = collect(1:m)
```
Comment (sbromberger) [inline src/randgraphs.jl]: Still - do you need this `collect`? Why won't the range do?

Comment (afternone) [inline src/randgraphs.jl]: targets is an array to store the target nodes. It will be updated in line 183, thus we need the collect.

Comment (sbromberger) [inline src/randgraphs.jl]: Ah, I see. Thanks.

**Comment by sbromberger** [comment]
PS: when you close these, it'd be really helpful to log a reason ("reopening with squashed commits")  or something :)

**Comment by afternone** [comment]
Good suggestion, I will pay an attention.

**Pull request closed without merging**